File: polyglot/csharp/signature_database_management.cs
using System;
using System.Collections.Generic;
using System.IO;
using System.Linq;

namespace polyglot.csharp
{
    public class SignatureDatabase
    {
        private readonly Dictionary<string, List<Signature>> _signaturesByExtension = new Dictionary<string, List<Signature>>();
        private readonly Dictionary<string, List<Signature>> _signaturesByMagic = new Dictionary<string, List<Signature>>();

        public void AddSignature(string extension, string magic, int length, string description)
        {
            var signature = new Signature(extension, magic, length, description);
            if (!_signaturesByExtension.ContainsKey(extension))
            {
                _signaturesByExtension[extension] = new List<Signature>();
            }
            _signaturesByExtension[extension].Add(signature);

            if (!_signaturesByMagic.ContainsKey(magic))
            {
                _signaturesByMagic[magic] = new List<Signature>();
            }
            _signaturesByMagic[magic].Add(signature);
        }

        public IEnumerable<Signature> FindByExtension(string extension)
        {
            if (_signaturesByExtension.TryGetValue(extension, out var signatures))
            {
                return signatures;
            }
            return Enumerable.Empty<Signature>();
        }

        public IEnumerable<Signature> FindByMagic(string magic)
        {
            if (_signaturesByMagic.TryGetValue(magic, out var signatures))
            {
                return signatures;
            }
            return Enumerable.Empty<Signature>();
        }

        public void LoadFromFile(string filePath)
        {
            if (!File.Exists(filePath))
            {
                throw new FileNotFoundException($"Signature file not found: {filePath}");
            }

            var lines = File.ReadAllLines(filePath);
            foreach (var line in lines)
            {
                var parts = line.Split(new[] { '\t' }, StringSplitOptions.RemoveEmptyEntries);
                if (parts.Length < 4)
                {
                    continue;
                }
                var extension = parts[0];
                var magic = parts[1];
                if (!int.TryParse(parts[2], out int length))
                {
                    continue;
                }
                var description = parts[3];
                AddSignature(extension, magic, length, description);
            }
        }

        public void SaveToFile(string filePath)
        {
            using (var writer = new StreamWriter(filePath))
            {
                foreach (var entry in _signaturesByExtension)
                {
                    foreach (var signature in entry.Value)
                    {
                        writer.WriteLine($"{signature.Extension}\t{signature.Magic}\t{signature.Length}\t{signature.Description}");
                    }
                }
            }
        }

        public class Signature
        {
            public string Extension { get; }
            public string Magic { get; }
            public int Length { get; }
            public string Description { get; }

            public Signature(string extension, string magic, int length, string description)
            {
                Extension = extension;
                Magic = magic;
                Length = length;
                Description = description;
            }
        }
    }

    class Program
    {
        static void Main(string[] args)
        {
            var db = new SignatureDatabase();

            // Example: Add some signatures
            db.AddSignature("txt", "ASCII", 1, "Text file");
            db.AddSignature("jpg", "JFIF", 4, "JPEG image");
            db.AddSignature("png", "PNG", 8, "Portable Network Graphics");

            Console.WriteLine("Loaded signatures:");
            foreach (var sig in db.FindByExtension("txt"))
            {
                Console.WriteLine($"- {sig.Description} ({sig.Magic}, length: {sig.Length})");
            }

            // Example: Save to file
            string tempFilePath = Path.GetTempFileName();
            db.SaveToFile(tempFilePath);
            Console.WriteLine($"Saved signature database to {tempFilePath}");

            // Example: Load from file
            var loadedDb = new SignatureDatabase();
            loadedDb.LoadFromFile(tempFilePath);
            Console.WriteLine("Loaded signatures from file:");
            foreach (var sig in loadedDb.FindByMagic("ASCII"))
            {
                Console.WriteLine($"- {sig.Description} ({sig.Extension}, length: {sig.Length})");
            }

            // Clean up
            File.Delete(tempFilePath);
        }
    }
}

File: polyglot/rust/signature_database_management.rs
use std::collections::{HashMap, HashSet};
use std::fs::File;
use std::io::{BufReader, BufRead};
use std::path::Path;

#[derive(Debug, Clone)]
pub struct Signature {
    pub name: String,
    pub pattern: Vec<u8>,
    pub offset: usize,
}

#[derive(Debug)]
pub struct SignatureDatabase {
    signatures: HashMap<String, Signature>,
    all_signatures: HashSet<String>,
}

impl SignatureDatabase {
    pub fn new() -> Self {
        SignatureDatabase {
            signatures: HashMap::new(),
            all_signatures: HashSet::new(),
        }
    }

    pub fn add_signature(&mut self, name: &str, pattern: &[u8], offset: usize) {
        let signature = Signature {
            name: name.to_string(),
            pattern: pattern.to_vec(),
            offset,
        };
        self.signatures.insert(name.to_string(), signature);
        self.all_signatures.insert(name.to_string());
    }

    pub fn load_from_file<P: AsRef<Path>>(&mut self, path: P) -> std::io::Result<()> {
        let file = File::open(path)?;
        let reader = BufReader::new(file);

        for line in reader.lines() {
            let line = line?;
            if line.trim().is_empty() {
                continue;
            }

            let parts: Vec<&str> = line.split_whitespace().collect();
            if parts.len() < 3 {
                continue;
            }

            let name = parts[0];
            let offset_str = parts[1];
            let pattern_str = parts[2..].join(" ");

            let offset = offset_str.parse::<usize>().ok()?;
            let pattern: Vec<u8> = pattern_str
                .as_bytes()
                .iter()
                .filter(|&&b| b != b' ')
                .map(|&b| b)
                .collect();

            self.add_signature(name, &pattern, offset);
        }

        Ok(())
    }

    pub fn get_signatures(&self) -> &HashMap<String, Signature> {
        &self.signatures
    }

    pub fn get_all_names(&self) -> &[String] {
        &self.all_signatures
    }
}

fn main() {
    let mut db = SignatureDatabase::new();
    // Example: Load signatures from a file
    if let Err(e) = db.load_from_file("signatures.txt") {
        eprintln!("Error loading signature database: {}", e);
        return;
    }

    println!("Loaded {} signatures:", db.get_all_names().len());
    for name in db.get_all_names() {
        println!("- {}", name);
    }

    // Example: Retrieve a signature by name
    if let Some(sig) = db.signatures.get("example_signature") {
        println!("Signature 'example_signature' has pattern: {:?}", sig.pattern);
        println!("Offset: {}", sig.offset);
    } else {
        println!("Signature 'example_signature' not found.");
    }
}